src/memoize-one.ts

```diff
@@ -1,40 +1,56 @@
 import areInputsEqual from './are-inputs-equal';
 
 // Using ReadonlyArray<T> rather than readonly T as it works with TS v3
-export type EqualityFn = (newArgs: any[], lastArgs: any[]) => boolean;
-
-function memoizeOne<
-  // Need to use 'any' rather than 'unknown' here as it has
-  // The correct Generic narrowing behaviour.
-  ResultFn extends (this: any, ...newArgs: any[]) => ReturnType<ResultFn>
->(resultFn: ResultFn, isEqual: EqualityFn = areInputsEqual): ResultFn {
-  let lastThis: unknown;
-  let lastArgs: unknown[] = [];
-  let lastResult: ReturnType<ResultFn>;
-  let calledOnce: boolean = false;
+export type EqualityFn<TFunc extends (...args: any[]) => any> = (
+  newArgs: Parameters<TFunc>,
+  lastArgs: Parameters<TFunc>,
+) => boolean;
+
+type Cache<TFunc extends (this: any, ...args: any[]) => any> = {
+  lastThis: ThisParameterType<TFunc>;
+  lastArgs: Parameters<TFunc>;
+  lastResult: ReturnType<TFunc>;
+};
+
+type MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> = {
```

Comment:
Previously we were casting the return value to be `ResultFn`

```ts
return memoized as ResultFn;
```

That is technically not correct as we are not maintaining any properties on the function itself. We now return a function that has the same type as the functions call signature and we are explicitly adding a `.clear` property to that